// discohook/src/classes/webhook.js
const fetch = require('node-fetch');
const send = require('../methods/send.js');
const edit = require('../methods/edit.js');
const dlt = require('../methods/delete.js');
const setUsername = require('../methods/setUsername.js');
const setAvatar = require('../methods/setAvatar.js');

module.exports = class Webhook {
    
    constructor(url) {
        if(typeof url !== 'string') throw new TypeError(`INVALID_WEBHOOK_STRING: No valid String has been provided. (WebHook class)`);
        this.webhook = fetch(url, { method: 'GET' }).then(res => res.json());
        this.url = url;
    };

    async send(options, webhookInfo) {
        const msg = await send(this.webhook, options, webhookInfo);
        return msg;
    };

    async edit(messageId, options, webhookInfo) {
        const editted = await edit(this.webhook, messageId, options, webhookInfo);
        return editted;
    };

    async delete(messageId) {
        await dlt(this.webhook, messageId);
    };

    async getInfo() {
        return (await this.webhook);
    };

    async setUsername(username) {
        await setUsername(this.webhook, username);
    };

    async setAvatar(avatar) {
        await setAvatar(this.webhook, avatar);
    };

};
