// discohook/src/classes/Builders/SelectMenuBuilder.js
module.exports = class SelectMenuBuilder {

    constructor() {
        this.menu = {
            type: 3
        };
    };

    setCustomId(id) {
        this.menu.custom_id = id;

        return this;
    };

    setPlaceholder(placeholder) {
        this.menu.placeholder = placeholder;

        return this;
    };

    setMinValues(value) {
        this.menu.min_values = value;

        return this;
    };

    setMaxValues(value) {
        this.menu.max_values = value;

        return this;
    };

    setDisabled(boolean) {
        this.menu.disabled = boolean;

        return this;
    };

    addOptions(options) {
        if(!Array.isArray(options) || options?.length <= 0 || options?.length > 25) throw new TypeError(`OPTIONS:INVALID_ARRAY: No valid Array has been provided or the length is higher than 25 or lower than 0.`);
        if(!this.menu.options) this.menu.options = [];

        for(const option of options) {
            this.menu.options.push(option);
        };

        return this;
    };

    toJSON() {
        return this.menu;
    };

};
